rmw_fastrtps_shared_cpp/src/rmw_take.cpp

```diff
@@ -76,6 +76,15 @@ _take(
   data.data = ros_message;
   data.impl = info->type_support_impl_;
   if (info->data_reader_->take_next_sample(&data, &sinfo) == ReturnCode_t::RETCODE_OK) {
+    if (subscription->options.ignore_local_publications) {
```

Comment:
@jamoralp meta: won't this result in waitsets waking up when there's no need to? Can we avoid that entirely?

Reply:
i am with @hidmic , i think this would affect performance. if there are many local publications, that means there are unnecessary executor loop to take & just drop the data? if i am not mistake.

Reply:
As the reader may have data queued from both local and remote publications, samples from local ones need to be consumed (i.e. taken). Fast DDS doesn't (yet) have a mechanism to filter samples before adding them to the history, which would be the ideal way of doing this. The same thing is done on [rmw_connextdds](https://github.com/ros2/rmw_connextdds/blob/721976a825360b260c2b8adc0e6a905a6ee508a9/rmw_connextdds_common/src/common/rmw_impl.cpp#L1643-L1654)

Reply:
I would make additional changes here, though.

First, `info->listener_->update_has_data(info->data_reader_)` should always be called after successfully taking a sample.
Second, I would wrap the code inside a `while (0 < info->data_reader_->get_unread_count())`, with a `break;` when we do `*taken = true;` and a `continue;` when a local publication sample is filtered.

Reply:
Changes addressed :+1:

Reply:
From [ros2/rmw_implementation (comment)](https://github.com/ros2/rmw_implementation/pull/192/files/254b69002e077fb9d9857f5c2bf7ee2a7a044274..f41148341892715da35bd7e0a33476539b1d4d8b#r646642889) I get that `rmw_cyclonedds_cpp` blocks message reception entirely when `ignore_local_publications` is set to true.

I'm not super thrilled about this variability. At the very least, I think it should be reflected in the `rmw` API contract for [ignore_local_publications](https://github.com/ros2/rmw/blob/a50f3bb1e24ac5338677990d74ab096a9d1776d7/rmw/include/rmw/types.h#L157-L170).

Reply:
It'd be good to address this ^ at some point.